In [1]:
pip install ipython-sql pymysql

Note: you may need to restart the kernel to use updated packages.


In [2]:
%load_ext sql
%config SqlMagic.style = '_DEPRECATED_DEFAULT'
%config SqlMagic.autopandas = True

In [ ]:
%sql mysql+pymysql://root:**********@localhost:3306/md_water_services

# Clustering data to unveil Maji Ndogo's water crisis
## Maji Ndogo: From analysis to action

In [8]:
%%sql 
SHOW TABLES;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
8 rows affected.


Tables_in_md_water_services
0             data_dictionary
1                    employee
2         global_water_access
3                    location
4                      visits
5               water_quality
6                water_source
7              well_pollution

In [9]:
%sql DESCRIBE data_dictionary;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


Field          Type Null Key Default Extra
0   table_name  varchar(255)   NO        None      
1  column_name  varchar(255)   NO        None      
2  description          text  YES        None      
3     datatype   varchar(50)  YES        None      
4   related_to  varchar(255)  YES        None

In [10]:
%%sql
SELECT*
FROM md_water_services.data_dictionary;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
49 rows affected.


table_name               column_name  \
0              employee      assigned_employee_id   
1              employee             employee_name   
2              employee              phone_number   
3              employee                     email   
4              employee                   address   
5              employee                 town_name   
6              employee             province_name   
7              employee                  position   
8                visits                 record_id   
9                visits               location_id   
10               visits                 source_id   
11               visits            time_of_record   
12               visits               visit_count   
13               visits             time_in_queue   
14               visits      assigned_employee_id   
15        water_quality                 record_id   
16        water_quality  subjective_quality_score   
17        water_quality               visit_count   
18         water_source                 source_id   
19         water_source      type_of_water_source   
20         water_source   number_of_people_served   
21       well_pollution                 source_id   
22       well_pollution                      date   
23       well_pollution               description   
24       well_pollution             pollutant_ppm   
25       well_pollution                biological   
26       well_pollution                   results   
27             location               location_id   
28             location                   address   
29             location             province_name   
30             location                 town_name   
31             location             location_type   
32  global_water_access                      name   
33  global_water_access                    region   
34  global_water_access                      year   
35  global_water_access                     pop_n   
36  global_water_access                     pop_u   
37  global_water_access                 wat_bas_n   
38  global_water_access                 wat_lim_n   
39  global_water_access               wat_unimp_n   
40  global_water_access                 wat_sur_n   
41  global_water_access                 wat_bas_r   
42  global_water_access                 wat_lim_r   
43  global_water_access               wat_unimp_r   
44  global_water_access                 wat_sur_r   
45  global_water_access                 wat_bas_u   
46  global_water_access                 wat_lim_u   
47  global_water_access               wat_unimp_u   
48  global_water_access                 wat_sur_u   

                                          description      datatype  \
0                 Unique ID assigned to each employee           INT   
1                                Name of the employee  VARCHAR(255)   
2                      Contact number of the employee   VARCHAR(15)   
3                       Email address of the employee  VARCHAR(255)   
4                 Residential address of the employee  VARCHAR(255)   
5         Name of the town where the employee resides  VARCHAR(255)   
6     Name of the province where the employee resides  VARCHAR(255)   
7               Position or job title of the employee  VARCHAR(255)   
8                    Unique ID assigned to each visit           int   
9                          ID of the location visited  varchar(255)   
10                     ID of the water source visited  varchar(510)   
11                         Date and time of the visit      datetime   
12             Number of visits made to this location           int   
13  Time spent by people waiting for water in a qu...           int   
14        ID of the employee who visited the location           int   
15                  Unique ID assigned to each record           int   
16  Score representing the subjective quality of t...           int   
17          Number of visits made for data collection           int   
18           

# DATA CLEANING

## Determining the names and emails of employees

In [14]:
%%sql
SELECT
employee_name,
LOWER(email) AS lower_email
FROM md_water_services.employee
LIMIT 20;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
20 rows affected.


employee_name lower_email
0      Amara Jengo        None
1      Bello Azibo        None
2     Bakari Iniko        None
3   Malachi Mavuso        None
4     Cheche Buhle        None
5   Zuriel Matembo        None
6     Deka Osumare        None
7   Lalitha Kaburi        None
8      Enitan Zuri        None
9        Farai Nia        None
10     Gamba Shani        None
11    Harith Nyota        None
12     Isoke Amani        None
13   Jengo Tumaini        None
14      Kunto Asha        None
15     Lesedi Kofi        None
16    Makena Thabo        None
17      Nia Furaha        None
18        Ona Sefu        None
19       Pili Zola        None

## Replacing trailing spaces with a fullstop

In [47]:
%%sql
SELECT
      CONCAT(
      LOWER(REPLACE(employee_name,' ','.')),'@endogowater.gov') AS new_email
FROM employee
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


new_email
0     amara.jengo@endogowater.gov
1     bello.azibo@endogowater.gov
2    bakari.iniko@endogowater.gov
3  malachi.mavuso@endogowater.gov
4    cheche.buhle@endogowater.gov
5  zuriel.matembo@endogowater.gov
6    deka.osumare@endogowater.gov
7  lalitha.kaburi@endogowater.gov
8     enitan.zuri@endogowater.gov
9       farai.nia@endogowater.gov

## Updating the Email Column

In [25]:
%%sql UPDATE md_water_services.employee
SET email = CONCAT(
          LOWER(REPLACE(employee_name,' ','.')),'@endogowater.gov');

 * mysql+pymysql://root:***@localhost:3306/md_water_services
56 rows affected.


Empty DataFrame
Columns: []
Index: []

In [48]:
%%sql
SELECT*
FROM employee
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


assigned_employee_id   employee_name  phone_number  \
0                     0     Amara Jengo  +99637993287   
1                     1     Bello Azibo  +99643864786   
2                     2    Bakari Iniko  +99222599041   
3                     3  Malachi Mavuso  +99945849900   
4                     4    Cheche Buhle  +99381679640   
5                     5  Zuriel Matembo  +99034075111   
6                     6    Deka Osumare  +99379364631   
7                     7  Lalitha Kaburi  +99681623240   
8                     8     Enitan Zuri  +99248509202   
9                    10       Farai Nia  +99570082739   

                            email                    address province_name  \
0     amara.jengo@endogowater.gov    36 Pwani Mchangani Road        Sokoto   
1     bello.azibo@endogowater.gov      129 Ziwa La Kioo Road      Kilimani   
2    bakari.iniko@endogowater.gov     18 Mlima Tazama Avenue       Hawassa   
3  malachi.mavuso@endogowater.gov         100 Mogadishu Road        Akatsi   
4    cheche.buhle@endogowater.gov           1 Savanna Street        Akatsi   
5  zuriel.matembo@endogowater.gov   26 Bahari Ya Faraja Road      Kilimani   
6    deka.osumare@endogowater.gov        104 Kenyatta Street        Akatsi   
7  lalitha.kaburi@endogowater.gov  145 Sungura Amanpour Road      Kilimani   
8     enitan.zuri@endogowater.gov           117 Kampala Road       Hawassa   
9       farai.nia@endogowater.gov  33 Angélique Kidjo Avenue        Amanzi   

  town_name        position  
0    Ilanga  Field Surveyor  
1     Rural  Field Surveyor  
2     Rural  Field Surveyor  
3    Lusaka  Field Surveyor  
4     Rural  Field Surveyor  
5     Rural  Field Surveyor  
6     Rural  Field Surveyor  
7     Rural  Field Surveyor  
8  Zanzibar  Field Surveyor  
9    Dahabu  Field Surveyor

In [49]:
%%sql
SELECT
LENGTH(phone_number)
FROM employee
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


LENGTH(phone_number)
0                    12
1                    12
2                    12
3                    12
4                    12
5                    12
6                    12
7                    12
8                    12
9                    12

## Removing Trailing Spaces in Phonenumbers

In [31]:
%%sql
SELECT
TRIM(phone_number)
FROM employee;

UPDATE employee
SET phone_number=TRIM(phone_number)

 * mysql+pymysql://root:***@localhost:3306/md_water_services
56 rows affected.
56 rows affected.


Empty DataFrame
Columns: []
Index: []

## Number of Employees in Each town

In [33]:
%%sql
SELECT
town_name,
COUNT(DISTINCT employee_name) AS num_employees
FROM employee
GROUP BY town_name;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
9 rows affected.


town_name  num_employees
0    Dahabu              6
1    Harare              5
2    Ilanga              3
3  Kintampo              1
4    Lusaka              4
5     Rural             29
6    Serowe              3
7   Yaounde              1
8  Zanzibar              4

## Total Number of records each employee Collected

In [50]:
%%sql
SELECT*
FROM visits
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


record_id location_id     source_id      time_of_record  visit_count  \
0          0   SoIl32582  SoIl32582224 2021-01-01 09:10:00            1   
1          1   KiRu28935  KiRu28935224 2021-01-01 09:17:00            1   
2          2   HaRu19752  HaRu19752224 2021-01-01 09:36:00            1   
3          3   AkLu01628  AkLu01628224 2021-01-01 09:53:00            1   
4          4   AkRu03357  AkRu03357224 2021-01-01 10:11:00            1   
5          5   KiRu29315  KiRu29315224 2021-01-01 10:17:00            1   
6          6   AkRu05234  AkRu05234224 2021-01-01 10:18:00            1   
7          7   KiRu28520  KiRu28520224 2021-01-01 10:28:00            1   
8          8   HaZa21742  HaZa21742224 2021-01-01 10:37:00            1   
9          9   AmDa12214  AmDa12214224 2021-01-01 10:58:00            1   

   time_in_queue  assigned_employee_id  
0             15                    12  
1              0                    46  
2             62                    40  
3              0                     1  
4             28                    14  
5              9                    40  
6              0                    30  
7              0                    34  
8              0                     6  
9              0                    36

In [126]:
%%sql
SELECT
assigned_employee_id,
COUNT(record_id) AS number_of_visits
FROM md_water_services.visits
GROUP BY assigned_employee_id
ORDER BY number_of_visits DESC
LIMIT 10;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


assigned_employee_id  number_of_visits
0                     1              3708
1                    30              3676
2                    34              3539
3                     3              3420
4                    10              3407
5                     8              3351
6                     5              3284
7                    36              3249
8                    48              2933
9                    28              2762

### Top employees contact details

In [125]:
%%sql
SELECT
employee_name,
email,
phone_number
FROM employee
WHERE assigned_employee_id IN(1,30,34)


 * mysql+pymysql://root:***@localhost:3306/md_water_services
3 rows affected.


employee_name                        email  phone_number
0   Bello Azibo  bello.azibo@endogowater.gov  +99643864786
1     Pili Zola    pili.zola@endogowater.gov  +99822478933
2    Rudo Imani   rudo.imani@endogowater.gov  +99046972648

## Analysing Locations

In [55]:
%%sql
SELECT*
FROM location
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


location_id               address province_name town_name location_type
0   AkHa00000    2 Addis Ababa Road        Akatsi    Harare         Urban
1   AkHa00001   10 Addis Ababa Road        Akatsi    Harare         Urban
2   AkHa00002    9 Addis Ababa Road        Akatsi    Harare         Urban
3   AkHa00003  139 Addis Ababa Road        Akatsi    Harare         Urban
4   AkHa00004   17 Addis Ababa Road        Akatsi    Harare         Urban
5   AkHa00005  125 Addis Ababa Road        Akatsi    Harare         Urban
6   AkHa00006   98 Addis Ababa Road        Akatsi    Harare         Urban
7   AkHa00007   21 Addis Ababa Road        Akatsi    Harare         Urban
8   AkHa00008   11 Addis Ababa Road        Akatsi    Harare         Urban
9   AkHa00009    6 Addis Ababa Road        Akatsi    Harare         Urban

In [60]:
%%sql
##Records per town
SELECT
town_name,
COUNT(town_name) AS Records_per_town
FROM location
GROUP BY town_name
ORDER BY Records_per_town DESC
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


town_name  Records_per_town
0     Rural             23740
1    Harare              1650
2     Amina              1090
3    Lusaka              1070
4    Mrembo               990
5    Asmara               930
6    Dahabu               930
7  Kintampo               780
8    Ilanga               780
9   Isiqalo               770

In [62]:
%%sql
SELECT
province_name,
COUNT(province_name) AS Records_per_province
FROM location
GROUP BY province_name
ORDER BY Records_per_province DESC;




 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


province_name  Records_per_province
0      Kilimani                  9510
1        Akatsi                  8940
2        Sokoto                  8220
3        Amanzi                  6950
4       Hawassa                  6030

In [66]:
%%sql
SELECT
province_name,
town_name,
COUNT(town_name) AS Records_per_town
FROM location
GROUP BY province_name, town_name
ORDER BY province_name, Records_per_town DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
31 rows affected.


province_name town_name  Records_per_town
0         Akatsi     Rural              6290
1         Akatsi    Lusaka              1070
2         Akatsi    Harare               800
3         Akatsi  Kintampo               780
4         Amanzi     Rural              3100
5         Amanzi    Asmara               930
6         Amanzi    Dahabu               930
7         Amanzi     Amina               670
8         Amanzi     Pwani               520
9         Amanzi   Abidjan               400
10        Amanzi     Bello               400
11       Hawassa     Rural              3900
12       Hawassa    Serowe               480
13       Hawassa     Amina               420
14       Hawassa      Deka               350
15       Hawassa    Djenne               310
16       Hawassa  Zanzibar               300
17       Hawassa   Yaounde               270
18      Kilimani     Rural              5440
19      Kilimani    Mrembo               990
20      Kilimani    Harare               850
21      Kilimani   Isiqalo               770
22      Kilimani      Zuri               750
23      Kilimani     Amara               710
24        Sokoto     Rural              5010
25        Sokoto    Ilanga               780
26        Sokoto   Majengo               620
27        Sokoto      Kofi               510
28        Sokoto    Bahari               470
29        Sokoto    Marang               460
30        Sokoto    Cheche               370

The above results show us that our field surveyors did an excellent job of documenting the status of our country's water crisis. Every province and town has many documented sources.

In [67]:
%%sql 
SELECT*
FROM water_source;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
39650 rows affected.


source_id type_of_water_source  number_of_people_served
0      AkHa00000224          tap_in_home                      956
1      AkHa00001224   tap_in_home_broken                      930
2      AkHa00002224   tap_in_home_broken                      486
3      AkHa00003224                 well                      364
4      AkHa00004224   tap_in_home_broken                      942
...             ...                  ...                      ...
39645  SoRu39645224          tap_in_home                      438
39646  SoRu39646224          tap_in_home                      614
39647  SoRu39647224                 well                      206
39648  SoRu39648224   tap_in_home_broken                      370
39649  SoRu39649224                 well                      202

[39650 rows x 3 columns]

In [69]:
%%sql
SELECT 
location_type,
COUNT(*) AS num_sources
FROM location
GROUP BY location_type;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
2 rows affected.
1 rows affected.


23740 / (15910 + 23740) * 100
0                       59.8739

In [72]:
%%sql
SELECT 
    location_type,
    COUNT(*) AS num_sources,
    COUNT(*) * 100.0 / SUM(COUNT(*)) OVER () AS pct
FROM location
GROUP BY location_type;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
2 rows affected.


location_type  num_sources       pct
0         Urban        15910  40.12610
1         Rural        23740  59.87390

60% of our water sources are in rural communities across Maji Ndogo.

## Diving into Sources

In [76]:
%%sql 
SELECT*
FROM water_source
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


source_id type_of_water_source  number_of_people_served
0  AkHa00000224          tap_in_home                      956
1  AkHa00001224   tap_in_home_broken                      930
2  AkHa00002224   tap_in_home_broken                      486
3  AkHa00003224                 well                      364
4  AkHa00004224   tap_in_home_broken                      942
5  AkHa00005224          tap_in_home                      736
6  AkHa00006224          tap_in_home                      882
7  AkHa00007224          tap_in_home                      554
8  AkHa00008224                 well                      398
9  AkHa00009224                 well                      346

### Total People Surveyed

In [78]:
%%sql 
SELECT 
SUM(number_of_people_served) As Total_people_surveyed
FROM water_source

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


Total_people_surveyed
0              27628140

#### How many of each of the different water source types there are

In [79]:
%%sql
SELECT
type_of_water_source,
COUNT(type_of_water_source) AS Number_water_sources
FROM water_source
GROUP BY type_of_water_source
ORDER BY Number_water_sources DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


type_of_water_source  Number_water_sources
0                 well                 17383
1          tap_in_home                  7265
2   tap_in_home_broken                  5856
3           shared_tap                  5767
4                river                  3379

It is clear that although there was a drought, water is still abundant in Maji Ndogo.

### Average number of people that are served by each water source?

In [81]:
%%sql
SELECT 
type_of_water_source,
ROUND(AVG(number_of_people_served),0) AS avg_people_served
FROM water_source
GROUP BY type_of_water_source;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


type_of_water_source avg_people_served
0          tap_in_home               644
1   tap_in_home_broken               649
2                 well               279
3           shared_tap              2071
4                river               699

The surveyors combined the data of many
households together and added this as a single tap record, but each household actually has its own tap. In addition to this, there is an average of
6 people living in a home. So 6 people actually share 1 tap (not 644).  
This means that 1 tap_in_home actually represents 644 ÷ 6 = ± 100 taps.  
Calculating the average number of people served by a single instance of each water source type helps us understand the typical capacity or load
on a single water source. This can help us decide which sources should be repaired or upgraded, based on the average impact of each upgrade.  
For example, wells don't seem to be a problem, as fewer people are sharing them.

### Total number of people served by each type of water source in total

In [84]:
%%sql
SELECT
type_of_water_source,
SUM(number_of_people_served) AS Total_people_served
FROM water_source
GROUP BY type_of_water_source
ORDER BY Total_people_served DESC;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


type_of_water_source Total_people_served
0           shared_tap            11945272
1                 well             4841724
2          tap_in_home             4678880
3   tap_in_home_broken             3799720
4                river             2362544

In [89]:
%%sql
SELECT
    type_of_water_source,
    SUM(number_of_people_served) AS Total_people_served,
    ROUND(
        SUM(number_of_people_served) * 100.0
        / SUM(SUM(number_of_people_served)) OVER (), 
        0
    ) AS percentage_people_per_source
FROM water_source
GROUP BY type_of_water_source
ORDER BY Total_people_served DESC;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


type_of_water_source Total_people_served percentage_people_per_source
0           shared_tap            11945272                           43
1                 well             4841724                           18
2          tap_in_home             4678880                           17
3   tap_in_home_broken             3799720                           14
4                river             2362544                            9

43% of our people are using shared taps in their communities, and on average, we saw earlier, that 2000 people share one shared_tap.
By adding tap_in_home and tap_in_home_broken together, we see that 31% of people have water infrastructure installed in their homes, but 45%
(14/31) of these taps are not working.  
This isn't the tap itself that is broken, but rather the infrastructure like treatment plants, reservoirs, pipes, and pumps that serve these homes that are broken.   
18% of people are using wells. But only 4916 out of 17383 are clean = 28% (from last week).

## SOLUTION

In [91]:
%%sql
SELECT 
    type_of_water_source,
    SUM(number_of_people_served) AS Total_people_served,
    RANK() OVER (
        ORDER BY SUM(number_of_people_served) DESC
    ) AS rank_by_population
FROM water_source
GROUP BY type_of_water_source
ORDER BY Total_people_served DESC;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


type_of_water_source Total_people_served  rank_by_population
0           shared_tap            11945272                   1
1                 well             4841724                   2
2          tap_in_home             4678880                   3
3   tap_in_home_broken             3799720                   4
4                river             2362544                   5

Ok, so we should fix shared taps first, then wells, and so on. But the next question is, which shared taps or wells should be fixed first?

In [97]:
%%sql
SELECT
source_id,
type_of_water_source,
number_of_people_served,
RANK() OVER(PARTITION BY type_of_water_source ORDER BY type_of_water_source DESC ) AS priority_rank
FROM water_source;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
39650 rows affected.


source_id type_of_water_source  number_of_people_served  \
0      SoRu37904224                river                      990   
1      SoRu38287224                river                      604   
2      SoRu38875224                river                      518   
3      SoRu37658224                river                      788   
4      SoRu39392224                river                      698   
...             ...                  ...                      ...   
39645  AmRu13224224                 well                      326   
39646  AmRu14060224                 well                      192   
39647  AmRu12885224                 well                      204   
39648  AmRu13740224                 well                      306   
39649  AmRu13935224                 well                      188   

       priority_rank  
0                  1  
1                  1  
2                  1  
3                  1  
4                  1  
...              ...  
39645              1  
39646              1  
39647              1  
39648              1  
39649              1  

[39650 rows x 4 columns]

By using RANK() teams doing the repairs can use the value of rank to measure how many they have fixed,

In [98]:
%%sql
SELECT
source_id,
type_of_water_source,
number_of_people_served,
DENSE_RANK() OVER(PARTITION BY type_of_water_source ORDER BY type_of_water_source DESC ) AS priority_rank
FROM water_source;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
39650 rows affected.


source_id type_of_water_source  number_of_people_served  \
0      SoRu37904224                river                      990   
1      SoRu38287224                river                      604   
2      SoRu38875224                river                      518   
3      SoRu37658224                river                      788   
4      SoRu39392224                river                      698   
...             ...                  ...                      ...   
39645  AmRu13224224                 well                      326   
39646  AmRu14060224                 well                      192   
39647  AmRu12885224                 well                      204   
39648  AmRu13740224                 well                      306   
39649  AmRu13935224                 well                      188   

       priority_rank  
0                  1  
1                  1  
2                  1  
3                  1  
4                  1  
...              ...  
39645              1  
39646              1  
39647              1  
39648              1  
39649              1  

[39650 rows x 4 columns]

# Analysing queues
The visits table documented all of the visits our field surveyors made to each location.  
For most sources, one visit was enough, but if there were queues, they visited the location a couple of times to get a good idea of the time it took for people to queue for water. So we have the time that they collected the data, how many times the site was visited, and how long people had to queue for water. 

In [99]:
%%sql
SELECT *
FROM visits 

 * mysql+pymysql://root:***@localhost:3306/md_water_services
60146 rows affected.


record_id location_id     source_id      time_of_record  visit_count  \
0              0   SoIl32582  SoIl32582224 2021-01-01 09:10:00            1   
1              1   KiRu28935  KiRu28935224 2021-01-01 09:17:00            1   
2              2   HaRu19752  HaRu19752224 2021-01-01 09:36:00            1   
3              3   AkLu01628  AkLu01628224 2021-01-01 09:53:00            1   
4              4   AkRu03357  AkRu03357224 2021-01-01 10:11:00            1   
...          ...         ...           ...                 ...          ...   
60141      60141   HaRu18629  HaRu18629224 2023-07-13 18:45:00            8   
60142      60142   KiRu26812  KiRu26812224 2023-07-14 09:36:00            8   
60143      60143   KiRu29092  KiRu29092224 2023-07-14 11:03:00            8   
60144      60144   AkRu02836  AkRu02836224 2023-07-14 13:05:00            8   
60145      60145   AkKi01525  AkKi01525224 2023-07-14 13:53:00            8   

       time_in_queue  assigned_employee_id  
0                 15                    12  
1                  0                    46  
2                 62                    40  
3                  0                     1  
4                 28                    14  
...              ...                   ...  
60141            170                    12  
60142             42                    10  
60143            187                    10  
60144             82                    12  
60145            323                    10  

[60146 rows x 7 columns]

In [101]:
%%sql
SELECT
MIN(time_of_record) AS Start_date,
MAX(time_of_record) AS End_date
FROM visits

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


Start_date            End_date
0 2021-01-01 09:10:00 2023-07-14 13:53:00

In [109]:
%%sql
SELECT 
    DATEDIFF('2023-07-14 13:53:00', '2021-01-01 09:10:00') AS Number_of_days;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


Number_of_days
0             924

## How long people have to queue on average in Maji Ndogo.

In [111]:
%%sql
SELECT 
    ROUND(AVG(NULLIF(time_in_queue, 0)),0) AS avg_queue_time_minutes
FROM visits;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


avg_queue_time_minutes
0                    123

So on average, people take two hours to fetch water if they don't have a tap in their homes.

## Queue times aggregated across the different days of the week.

In [112]:
%%sql
SELECT
    DAYNAME(time_of_record) AS day_of_week,
    ROUND(AVG(NULLIF(time_in_queue, 0))) AS avg_queue_time_minutes
FROM visits
GROUP BY day_of_week
ORDER BY FIELD(day_of_week, 'Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday');


 * mysql+pymysql://root:***@localhost:3306/md_water_services
7 rows affected.


day_of_week avg_queue_time_minutes
0      Monday                    137
1     Tuesday                    108
2   Wednesday                     97
3    Thursday                    105
4      Friday                    120
5    Saturday                    246
6      Sunday                     82

Saturdays have much longer queue times compared to the other days!

## Average queue time by hour of the day

In [113]:
%%sql
SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
    ROUND(AVG(NULLIF(time_in_queue, 0))) AS avg_queue_time_minutes
FROM visits
GROUP BY hour_of_day
ORDER BY hour_of_day;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
14 rows affected.


hour_of_day avg_queue_time_minutes
0        06:00                    149
1        07:00                    149
2        08:00                    149
3        09:00                    118
4        10:00                    114
5        11:00                    111
6        12:00                    112
7        13:00                    115
8        14:00                    114
9        15:00                    114
10       16:00                    114
11       17:00                    149
12       18:00                    147
13       19:00                    168

Mornings and evenings are the busiest. It looks like people collect water before and after work.

# Reporting Insights

In [115]:
%%sql
SELECT
      TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
      DAYNAME(time_of_record) AS Day,
CASE
    WHEN DAYNAME(time_of_record) = 'Sunday' THEN time_in_queue
    ELSE NULL
END AS Sunday
FROM visits
WHERE
time_in_queue != 0; 

 * mysql+pymysql://root:***@localhost:3306/md_water_services
29642 rows affected.


hour_of_day       Day  Sunday
0           09:00    Friday     NaN
1           09:00    Friday     NaN
2           10:00    Friday     NaN
3           10:00    Friday     NaN
4           11:00    Friday     NaN
...           ...       ...     ...
29637       18:00  Thursday     NaN
29638       09:00    Friday     NaN
29639       11:00    Friday     NaN
29640       13:00    Friday     NaN
29641       13:00    Friday     NaN

[29642 rows x 3 columns]

In [116]:
%%sql 
SELECT
     TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,-- Sunday
    ROUND(AVG(
CASE
    WHEN DAYNAME(time_of_record) = 'Sunday' THEN time_in_queue
ELSE NULL
END
),0) AS Sunday,
ROUND(AVG(CASE
         WHEN DAYNAME(time_of_record) = 'Monday' THEN time_in_queue
ELSE NULL
END
),0) AS Monday,
ROUND(AVG(CASE
         WHEN DAYNAME(time_of_record) = 'Tuesday' THEN time_in_queue
ELSE NULL
END
),0) AS Tuesday,
ROUND(AVG(CASE
         WHEN DAYNAME(time_of_record) = 'Wednesday' THEN time_in_queue
ELSE NULL
END
),0) AS Wednesday,
ROUND(AVG(CASE
         WHEN DAYNAME(time_of_record) = 'Thursday' THEN time_in_queue
ELSE NULL
END
),0) AS Thursday,
ROUND(AVG(CASE
         WHEN DAYNAME(time_of_record) = 'Friday' THEN time_in_queue
ELSE NULL
END
),0) AS Friday,
ROUND(AVG(CASE
         WHEN DAYNAME(time_of_record) = 'Saturday' THEN time_in_queue
ELSE NULL
END
),0) AS Saturday

FROM visits
WHERE time_in_queue != 0
GROUP BY hour_of_day
ORDER BY hour_of_day;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
14 rows affected.


hour_of_day Sunday Monday Tuesday Wednesday Thursday Friday Saturday
0        06:00     79    190     134       112      134    153      247
1        07:00     82    186     128       111      139    156      247
2        08:00     86    183     130       119      129    153      247
3        09:00     84    127     105        94       99    107      252
4        10:00     83    119      99        89       95    112      259
5        11:00     78    115     102        86       99    104      236
6        12:00     78    115      97        88       96    109      239
7        13:00     81    122      97        98      101    115      242
8        14:00     83    127     104        92       96    110      244
9        15:00     83    126     104        88       92    110      248
10       16:00     83    127      99        90       99    109      251
11       17:00     79    181     135       121      129    151      251
12       18:00     80    174     122       113      132    158      240
13       19:00    127    159     145       176      137    103      282

1. Queues are very long on a Monday morning and Monday evening as people rush to get water.
2. Wednesday has the lowest queue times, but long queues on Wednesday evening.
3. People have to queue pretty much twice as long on Saturdays compared to the weekdays. It looks like people spend their Saturdays queueing
for water, perhaps for the week's supply.
4. The shortest queues are on Sundays, and this is a cultural thing. The people of Maji Ndogo prioritise family and religion, so Sundays are spent
with family and friends.

# Water Accessibility and infrastructure summary report
This survey aimed to identify the water sources people use and determine both the total and average number of users for each source.
Additionally, it examined the duration citizens typically spend in queues to access water.

## Insights
1. Most water sources are rural.
2. 43% of our people are using shared taps. 2000 people often share one tap.
3. 31% of our population has water infrastructure in their homes, but within that group, 45% face non-functional systems due to issues with pipes,
pumps, and reservoirs.
4. 18% of our people are using wells of which, but within that, only 28% are clean..
5. Our citizens often face long wait times for water, averaging more than 120 minutes.
6. In terms of queues:
- Queues are very long on Saturdays.
- Queues are longer in the mornings and evenings.
- Wednesdays and Sundays have the shortest queues.

## Our Plan

1. We want to focus our efforts on improving the water sources that affect the most people.
- Most people will benefit if we improve the shared taps first.
- Wells are a good source of water, but many are contaminated. Fixing this will benefit a lot of people.
- Fixing existing infrastructure will help many people. If they have running water again, they won't have to queue, thereby shorting queue times for
others. So we can solve two problems at once.
- Installing taps in homes will stretch our resources too thin, so for now, if the queue times are low, we won't improve that source.
2. Most water sources are in rural areas. We need to ensure our teams know this as this means they will have to make these repairs/upgrades in
rural areas where road conditions, supplies, and labour are harder challenges to overcome.

# Practical solutions
1. If communities are using rivers, we can dispatch trucks to those regions to provide water temporarily in the short term, while we send out
crews to drill for wells, providing a more permanent solution.
2. If communities are using wells, we can install filters to purify the water. For wells with biological contamination, we can install UV filters that
kill microorganisms, and for *polluted wells*, we can install reverse osmosis filters. In the long term, we need to figure out why these sources
are polluted.
3. For shared taps, in the short term, we can send additional water tankers to the busiest taps, on the busiest days. We can use the queue time
pivot table we made to send tankers at the busiest times. Meanwhile, we can start the work on installing extra taps where they are needed.
According to UN standards, the maximum acceptable wait time for water is 30 minutes. With this in mind, our aim is to install taps to get
queue times below 30 min.
4. Shared taps with short queue times (< 30 min) represent a logistical challenge to further reduce waiting times. The most effective solution,
installing taps in homes, is resource-intensive and better suited as a long-term goal.
5. Addressing broken infrastructure offers a significant impact even with just a single intervention. It is expensive to fix, but so many people
can benefit from repairing one facility. For example, fixing a reservoir or pipe that multiple taps are connected to. We will have to find the
commonly affected areas though to see where the problem actually is.